src/worker.rs

```diff
@@ -0,0 +1,320 @@
+    /// business logic as soon as you consider the worker is initialized,
+    /// otherwise the parent supervisor will block and eventually fail with a
+    /// timeout.
+    ///
+    /// ### Report a start error with the StartNotifier
+    ///
+    /// If for some reason, a worker is not able to start correctly (e.g. DB
+    /// connection fails, network is kaput, etc.), the worker may call the
+    /// `StartNotifier#failed` function with the impending error as a parameter.
+    /// This will cause the whole supervision system start procedure to abort
+    /// and fail fast.
+    ///
+    pub fn new_with_start<F, O>(name: &str, mut routine0: F) -> Self
+    where
+        F: FnMut(Context, StartNotifier) -> O + 'static,
+        O: Future<Output = anyhow::Result<()>> + FutureExt + Send + Sized + 'static,
+    {
+        let routine1 = move |ctx: Context, on_start: StartNotifier| routine0(ctx, on_start).boxed();
+        Spec {
+            name: name.to_owned(),
+            shutdown: Shutdown::Indefinitely,
+            restart: Restart::Permanent,
+            routine: Box::new(routine1),
+        }
+    }
+
+    /// restart is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn restart(mut self, restart: Restart) -> Self {
+        self.restart = restart;
+        self
+    }
+
+    /// shutdown is a builder method that allows to modify the Restart settings
+    /// for the spawned worker
+    pub fn shutdown(mut self, shutdown: Shutdown) -> Self {
+        self.shutdown = shutdown;
+        self
+    }
+
+    /// start spawns a new routine that executes the start function of this
+    /// Spec.
+    ///
+    /// ### Blocking on start function
+    ///
+    /// This function will block and wait until a notification from the
+    /// start function (executing on a newly spawned routine) is received; this
+    /// notification indicates if the start was successful or not.
+    ///
+    /// This blocking is necessary to ensure that supervised worker trees get
+    /// spawned in the specified/expected order.
+    pub async fn start(
+        mut self,
+        parent_ctx: &Context,
+        parent_name: &str,
+    ) -> anyhow::Result<Worker> {
+        let runtime_name = format!("{}/{}", parent_name, self.name);
+        let created_at = Utc::now();
+
+        // ON_START setup
+        let (started_tx, started_rx) = oneshot::channel::<Option<anyhow::Error>>();
+        let start_notifier = StartNotifier::from_oneshot(started_tx);
+
+        // CANCEL setup
+        let (ctx, termination_handle) = Context::with_cancel(parent_ctx);
+        let (kill_handle, kill_registration) = AbortHandle::new_pair();
+        let future = (*self.routine)(ctx, start_notifier);
+        let routine = Abortable::new(future, kill_registration);
+
+        // SPAWN -- actually spawn concurrent worker future
+        let join_handle = task::spawn(routine);
```

Comment:
Is there a reason not to model the startup completion as a `Future` directly? Then if the client code wants to, it can use this same oneshot pattern to let the spawned task run fully independently and do setup as part of the same body of code as the work. This oneshot notification API feels more error prone to use.  I think it will be a bit more complex to implement that way though and it might make sense to introduce a trait that has a pair of functions something like this:

```rust
pub trait Worker {
    type Error: std::error::Error;

    fn start(&Context) -> BoxFuture<Result<Self, Self::Error>>;
    fn run(&mut self, &Context) -> BoxFuture<Result<(), Self::Error>;
}
```

Reply:
I don't necessarily think that this change needs to be made right away, just curious if you had thought about that representation.

Reply:
I have considered it, but I came up with some concerns that might be unfounded due to my lack of Rust skills:

* If we provide a Worker `trait`, then we need to increase the bootstrap footprint (code line wise), define a `type struct` and  an `impl for` block. Not sure if that footprint would help or get in the way, as I need to come up with names for those structs.

* As soon as we have a `Worker` trait, we have an implementing struct that needs to manage state in a thread-safe manner. On the other hand, if we `move` the state to a closure, things become simpler IMO. That is a matter of opinion and I could be wrong if the user experience in Rust end up being worst this way (I'm lacking the experience to know for sure)

* Rust quesiton: Is the `BoxFuture` return value necessary on the trait? I would prefer return a `impl Future` and do the boxing on the internal API (no need to make our client call `boxed()` every time.

* Rust question: can I have a default implementation for the `start` trait function?

* Rust question: What is the benefit of having an associated `type Error` to the trait?

Reply:
> * If we provide a Worker `trait`, then we need to increase the bootstrap footprint (code line wise), define a `type struct` and  an `impl for` block. Not sure if that footprint would help or get in the way, as I need to come up with names for those structs.

Yeah, agreed there is for sure more lines of code and it's (at least slightly) more complex code.

> * As soon as we have a `Worker` trait, we have an implementing struct that needs to manage state in a thread-safe manner. On the other hand, if we `move` the state to a closure, things become simpler IMO. That is a matter of opinion and I could be wrong if the user experience in Rust end up being worst this way (I'm lacking the experience to know for sure)

I'm not sure this is true. I made several potentially questionable choices in my suggested trait that are not required. In particular, having `start` return a `Self` type on success, and then having `run` take a `&self`. Both of those could go away, and then the trait would basically just be a container for pointers to two different functions.

```rust
pub trait WorkerRunner {
    fn start(&Context) -> Result<(), Error> { Ok(()) } // default no-op implementation
    fn run(&Context) -> BoxFuture<Result<(), Error>>;
}
```

This is basically a way of tying together the two boxed functions that get passed to `start` and `run`.

I should note that this change would make far more sense with something I didn't explicitly mention, which is that I don't think the worker future should be `spawned` at all necessarily.  Using `spawn` enforces a lot of constraints on the Futures that can be created since they need to be `Future + Send + 'static`. If you don't spawn, and instead just await the `start` future directly inside of the `start` function (with timeout of course) then you don't have to require those bounds necessarily.  And like I said in my first comment on this thread, this `spawn` and communicate with the supervisor through channels can then still be implemented by a user of the library, but it isn't enforced by the API.  The real motivation for this comment was the documentation saying that it's important to remember to send the `started` notification, and I thought "why not just have that requirement be part of the type structure?"

Ultimately though, I'm not positive that my suggested design is better than what you have so far, especially since I'm only vaguely familiar with the supervisor tree pattern. It seems to me though that a supervisor needs to be a future itself, and that a supervisor typically only watches one future, so it's future implementation could just be delegating to it's worker future and then handling it's return type using appropriate logic. Getting rid of the `spawn` also gets rid of the need for any `Worker` struct implementing the `Worker` trait to even be thread-aware because the supervisor future would always actually just be composed with the worker it is supervising.

> * Rust quesiton: Is the `BoxFuture` return value necessary on the trait? I would prefer return a `impl Future` and do the boxing on the internal API (no need to make our client call `boxed()` every time.

Unfortunately yes, currently `impl` trait return types are not supported in traits. Here's [the rationale for why it wasn't targeted in the original impl trait RFC](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md#limitation-to-freeinherent-functions).

Ideally, I think that both `start` and `run` would actually be `async` functions, but those are also not directly supported in Rust currently (though there is a library that implements it as a macro [`async-trait`](https://lib.rs/crates/async-trait)). Using that library might be very reasonable in this case though, because then we'd get the best of both worlds, don't require the user to box all their returned futures, let them write an `async` function instead of wrapping an async function in a boilerplate trait implementation that calls the `async` function and then boxes the returned future, and it's easier to use the client API from the code in this library. 

> * Rust question: can I have a default implementation for the `start` trait function?

Yes! Just provide the implementation in the trait definition instead of ending with a semicolon. It can still be overriden by trait implementors.

> * Rust question: What is the benefit of having an associated `type Error` to the trait?

It's essentially a generic parameter to the trait. This lets the user specify what type of error their worker's create, so we aren't forcing them to box the error into a trait object. If we use the `std::error::Error` trait bound on it then we can always do the boxing in the calling context if it needs to be propagated, but I think most of the strategies would just log the error in some way, so then the boxing isn't necessary (correct me if I'm wrong on this assumption).

Using an associated type instead of a `trait Worker<Error>` makes it so that any type can only implement `Worker` once, instead of "once for each error type they specify" which would tend to get weird.

Reply:
> I don't think the worker future should be spawned at all necessarily. Using spawn enforces a lot of constraints on the Futures that can be created since they need to be Future + Send + 'static

Oh this interesting, so I had the impression that spawn would guarantee some sort of green thread on the tokio runtime, this may not be true? Is it like a superfluous call?

>  And like I said in my first comment on this thread, this spawn and communicate with the supervisor through channels can then still be implemented by a user of the library, but it isn't enforced by the API.

Not sure if I'm understanding this line correctly. It was my purpose to enforce the start order (internally) to guarantee that one worker gets started before another. The OTP API allow you to decide if worker routines of a supervisor get spawned from `LeftToRight` or `RightToLeft`, and this `oneshot` pattern plays a role there.

> I thought "why not just have that requirement be part of the type structure?"

I like reviewers asking that question, thank you for that.

>  It seems to me though that a supervisor needs to be a future itself, and that a supervisor typically only watches one future

The supervisor is a task, yes, although it may (and on this implementation will) supervise more than worker.

> Using an associated type instead of a trait Worker<Error> makes it so that any type can only implement Worker once, instead of "once for each error type they specify" which would tend to get weird.

My experience with adding a type parameter to Worker types is that later it becomes more difficult to compose workers in HashMap data structures. 

We will need to have collections of workers on the supervisor code, and if we can monomorphize the error type, the better. 

In theory we do not need to know the error, as the de facto response is going to be to restart or fail hard. Maybe we could offer the boxed error on an API callback so that API clients can make a more smart decision about restart logic depending on the error, but that breads complexity, so not so sure is the right way to go.

---

After reading the design decisions here, it sounds to me that the tradeoff balance goes towards the current way of things, with a chance to evolve towards the trait approach once they play better with `async` (e.g. support of `async fn` or return of `impl Future` in traits).

The things I really do not like:

* Req of a `.boxed` call on return of a future
* Req of yet another heavy third party library (async-trait) into someones project.

Calling the `StartNotifier` is not _terrible_, as we will get a warning if you do not call the `StartNotifier` parameter, and if you don't really need it, you are not going to call `new_with_start`. Also, you'll get a fast/hard runtime error on start if you do not call it (see the new test [`test_worker_start_not_invoked`](https://github.com/capatazlib/rust-capataz/pull/3/files#diff-59ff9455670b6e0eeceff9e934b91c8aR396))

Reply:
> Oh this interesting, so I had the impression that spawn would guarantee some sort of green thread on the tokio runtime, this may not be true? Is it like a superfluous call?

It creates a task in the runtime which is sort of analogous to a green thread yes. Since the supervisor is going to be watching multiple workers then this is the right thing to do I think. Whether it can actually be run in parallel is up to whether the runtime has multiple threads or not.  Because it could end up running a different thread than the one that called `spawn` is the reason that the `'static + Send` bounds are needed.

> Not sure if I'm understanding this line correctly. It was my purpose to enforce the start order (internally) to guarantee that one worker gets started before another. The OTP API allow you to decide if worker routines of a supervisor get spawned from LeftToRight or RightToLeft, and this oneshot pattern plays a role there.

I think this ordering enforcement is still possible with the trait implementation I was thinking of, but since I think other design pressures are pointing against using it, it's not necessary to delve more deeply into the details.

> My experience with adding a type parameter to Worker types is that later it becomes more difficult to compose workers in HashMap data structures.

A very good point. Using a boxed error for that purpose definitely makes sense. I think using a boxed `std::error::Error` is probably the right move as a library, even though that's basically what `anyhow` provides, in order to avoid imposing the use of anyhow in client code that may not want it. Though perhaps that is better done as a refinement before release (or perhaps not at all if a compelling reason to keep it comes up).

> The things I really do not like:
>
>    Req of a .boxed call on return of a future
>    Req of yet another heavy third party library (async-trait) into someones project.
>
> Calling the StartNotifier is not terrible, as we will get a warning if you do not use the StartNotifier parameter, and if you don't really need it, you are not going to call new_with_start. Also, you'll get a fast/hard runtime error on start if you do not call it (see the new test test_worker_start_not_invoked)

Yeah, all fair issues with my alternative.

> After reading the design decisions here, it sounds to me that the tradeoff balance goes towards the current way of things, with a chance to evolve towards the trait approach once they play better with async (e.g. support of async fn or return of impl Future in traits).

I agree. Unfortunately it will probably be quite a while before traits get support for `impl` return types or direct async method support. And probably it won't be very necessary for anyone to get the level of control that I'm thinking of.  I will be interested to see how client code for this API ends up looking!

Reply:
Cool beans, Is it ok if I marked this as resolved?

Reply:
yup!